src/Common/tests/Collections/IEnumerableTest.cs

```diff
@@ -0,0 +1,428 @@
+                                   for (int i = 0; i < 3; i++)
+                                   {
+                                       Assert.Throws<NotSupportedException>(
+                                           () => enumerator.Reset());
+                                   }
+                                   VerifyEnumerator(
+                                       items,
+                                       enumerator,
+                                       items.Length/2,
+                                       items.Length - (items.Length/2),
+                                       true,
+                                       false);
+                               }
+                               else if (iter == 2)
+                               {
+                                   VerifyEnumerator(items, enumerator);
+                                   for (int i = 0; i < 3; i++)
+                                   {
+                                       Assert.Throws<NotSupportedException>(
+                                           () => enumerator.Reset());
+                                   }
+                                   VerifyEnumerator(items, enumerator, 0, 0, true, false);
+                               }
+                               else
+                               {
+                                   VerifyEnumerator(items, enumerator);
+                               }
+                           });
+            }
+            else
+            {
+                RepeatTest((enumerator, items, iter) =>
+                           {
+                               if (iter == 1)
+                               {
+                                   VerifyEnumerator(items, enumerator, 0, items.Length / 2, false, true);
+                                   enumerator.Reset();
+                                   enumerator.Reset();
+                               }
+                               else if (iter == 3)
+                               {
+                                   VerifyEnumerator(items, enumerator);
+                                   enumerator.Reset();
+                                   enumerator.Reset();
+                               }
+                               else
+                               {
+                                   VerifyEnumerator(items, enumerator);
+                               }
+                           }, 5);
+            }
+        }
+
+        [Fact]
+        public void ModifyCollectionWithNewEnumerator()
+        {
+            var enumerable = GetEnumerable(GenerateItems(16));
+            var enumerator = enumerable.GetEnumerator();
+            InvalidateEnumerator(enumerable);
+            VerifyModifiedEnumerator(enumerator, null, true, false);
+        }
+
+        [Fact]
+        public void EnumerateFirstItemThenModify()
+        {
+            var items = GenerateItems(16);
```

Comment:
Is 16 items special? You seem to be making them a lot.  I'm wondering if it warrants a const